Add checks for the case of data url in SVGURIReference.

ziransun opened the pull request:
#### 04332cec0029ebb437fefa67af8e4e10b2f72ad9
<pre>
Add checks for the case of data url in SVGURIReference.
<a href="https://bugs.webkit.org/show_bug.cgi?id=149460">https://bugs.webkit.org/show_bug.cgi?id=149460</a>

Reviewed by Said Abou-Hallawa.

Add check for the case of data url in SVGURIReference.
The test still fails as it returns nullptr for the element in data url case in
SVGURIReference::targetElementFromIRIString(). We need to load the data url in
a Document to be able to get the target element for the corresponding fragement
ID.

* LayoutTests/TestExpectations:
* Source/WebCore/svg/SVGURIReference.cpp:
(WebCore::SVGURIReference::targetElementFromIRIString):
(WebCore::SVGURIReference::haveLoadedRequiredResources const):
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::updateExternalDocument):

Canonical link: <a href="https://commits.webkit.org/285321@main">https://commits.webkit.org/285321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7072bc48039fb18886169b4052de084285a70d7b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72023 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24808 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76183 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23232 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74138 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59244 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23052 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56823 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15322 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75090 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46643 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62040 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37258 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43306 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19512 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21578 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65214 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19874 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77865 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16263 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19045 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65302 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16309 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62063 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64555 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15956 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12743 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6389 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47241 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2025 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48310 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49597 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/SVGURIReference.h:
```diff
@@ -54,7 +54,6 @@ class SVGURIReference {
 
         // If the URI matches our documents URL, we're dealing with a local reference.
         URL url = document.completeURL(uri);
-        ASSERT(!url.protocolIsData());
```
What is the goal of this change?

Removing this assertion will prevent a crash in the debug build. But it will not affect the release build. More importantly this change will not fix the actual bug which can happen in cases like imported/w3c/web-platform-tests/css/filter-effects/svg-feimage-002.html.

The code in `SVGURIReference::targetElementFromIRIString()` assumes the `targetElement` is either local in the current `treeScope` or it is an element inside an `externalDocument`. It does not handle the case when the URL is data url followed by the fragment ID. So we end up making this early return:
```
    // Exit early if the referenced url is external, and we have no externalDocument given.
    if (isExternalURIReference(iri, document))
        return { nullptr, WTFMove(id) };
```
I think fixing the actual bug is more important than removing an assertion which reminds us that our code is not complete yet.

ziransun replied:
Updated but FE test is not fixed yet. I wonder if we could file an different bug on that. PTAL. Thank you!

shallawa commented on Source/WebCore/svg/SVGURIReference.cpp:
```diff
@@ -102,6 +102,13 @@ auto SVGURIReference::targetElementFromIRIString(const String& iri, const TreeSc
 
+    if (url.protocolIsData()) {
+        // We need to get the actual element that corresponds to the fragment id.
+        // TODO: We might need to add a parser to parse data to an actual documentfragment in order to get the element in the documentfragment.
```
We usually use FIXME instead of TODO.

shallawa replied:
Also the comment should say something like 
> FIXME: We need to load the data url in a Document to able to get the target element.

We can't limit the loader to load a documentfragment. We have to load everything. And we do not need to add a parser. We already have a mechanism to load a data url in a Document and it is used in other places like 
```
.box {
    background: url('data:image/svg+xml;utf8,<svg ...> ... </svg>');
}
```
and
```
<img src='data:image/svg+xml;utf8,<svg ... > ... </svg>'>
```

ziransun replied:
Thanks very much for the explanation!

shallawa commented on Source/WebCore/svg/SVGUseElement.cpp:
```diff
@@ -618,6 +618,8 @@ void SVGUseElement::updateExternalDocument()
 {
     URL externalDocumentURL;
     Ref<Document> document = this->document();
+    if (document->completeURL(href()).protocolIsData())
+        return;
```
I think we should add a FIXME comment here as well. The data url should be treated as an external URL and we should create an external Document for it. We can add something like:

> FIXME: This early exit should be removed once the ASSERT(!url.protocolIsData()) is removed from isExternalURIReference()

ziransun committed:
Add checks for the case of data url in SVGURIReference.
https://bugs.webkit.org/show_bug.cgi?id=149460

Reviewed by Said Abou-Hallawa.

Add check for the case of data url in SVGURIReference.
The test still fails as it returns nullptr for the element in data url case in
SVGURIReference::targetElementFromIRIString(). We need to load the data url in
a Document to be able to get the target element for the corresponding fragement
ID.

* LayoutTests/TestExpectations:
* Source/WebCore/svg/SVGURIReference.cpp:
(WebCore::SVGURIReference::targetElementFromIRIString):
(WebCore::SVGURIReference::haveLoadedRequiredResources const):
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::updateExternalDocument):

Canonical link: https://commits.webkit.org/285321@main
